lib/src/protocol/kawa_h1/mod.rs

```diff
@@ -0,0 +1,1933 @@
+            size
+        );
+
+        if size > 0 {
+            self.response_stream.storage.fill(size);
+            count!("back_bytes_in", size as i64);
+            metrics.backend_bin += size;
+            // if self.kawa_response.storage.is_full() {
+            //     self.backend_readiness.interest.remove(Ready::READABLE);
+            // }
+        } else {
+            self.backend_readiness.event.remove(Ready::READABLE);
+        }
+
+        // TODO: close delimited and backend_hup should be handled better
+        match socket_state {
+            SocketResult::Error => {
+                backend_socket.read_error();
+                self.log_request_error(
+                    metrics,
+                    &format!(
+                        "back socket {socket_state:?}, closing session.  Readiness: {:?} -> {:?}, read {size} bytes",
+                        self.frontend_readiness,
+                        self.backend_readiness,
+                    ),
+                );
+                return SessionResult::Close;
+            }
+            SocketResult::WouldBlock | SocketResult::Closed => {
+                self.backend_readiness.event.remove(Ready::READABLE);
+            }
+            SocketResult::Continue => {}
+        }
+
+        self.backend_readable_parse(metrics)
+    }
+
+    pub fn backend_readable_parse(&mut self, metrics: &mut SessionMetrics) -> SessionResult {
+        println!("==============backend_readable_parse");
+        let was_initial = self.response_stream.is_initial();
+
+        kawa::h1::parse(&mut self.response_stream, &mut self.context);
+        kawa::debug_kawa(&self.response_stream);
+
+        if self.response_stream.is_error() {
+            incr!("http.backend_parse_errors");
+            // was_initial is maybe too restrictive
+            // from what I understand the only condition should be:
+            // "did we already sent byte?"
+            if was_initial {
+                self.set_answer(DefaultAnswerStatus::Answer502, None);
+                return SessionResult::Continue;
+            } else {
+                return SessionResult::Close;
+            }
+        }
+
+        if self.response_stream.is_main_phase() {
+            self.frontend_readiness.interest.insert(Ready::WRITABLE);
+        }
+        if self.response_stream.is_terminated() {
+            metrics.backend_stop();
+            self.backend_stop = Some(Instant::now());
+            self.backend_readiness.interest.remove(Ready::READABLE);
+        }
+        SessionResult::Continue
+    }
+}
+
+impl<Front: SocketHandler, L: ListenerHandler + L7ListenerHandler> Http<Front, L> {
+    fn log_context(&self) -> LogContext {
+        LogContext {
+            request_id: self.context.id,
+            cluster_id: self.cluster_id.as_deref(),
+            backend_id: self.backend_id.as_deref(),
+        }
+    }
+
+    /// (code, reason)
+    pub fn get_response_line(&self) -> (Option<u16>, Option<&str>) {
```

Comment:
please write more explicit comments like "parses HTTP response code and reason from the status line"